File: apps/api/app/core/audio_stream_crypto.py
"""
AES-256-GCM decryption for /ws/voice-stream's audio chunks (spec: encrypt
on-device, decrypt server-side, so raw microphone audio is never sent in
the clear over the wire — see VoiceClassifierClient.kt's sendAudioChunk
for the encrypting side).

Wire format (matches the Kotlin side byte-for-byte): the client sends
base64(nonce[12 bytes] || ciphertext || tag[16 bytes]) as `audio_chunk_b64`
— AES-GCM's standard 12-byte nonce prepended to its own authenticated
ciphertext, one field, no separate nonce parameter needed.

Key: settings.voice_stream_audio_key_b64, a static pre-shared 32-byte key
(see config.py's own caveat on this — not a per-session negotiated key,
a stated scope limitation for this project's current stage).
"""

import base64
import logging

from cryptography.exceptions import InvalidTag
from cryptography.hazmat.primitives.ciphers.aead import AESGCM

from app.core.config import settings

logger = logging.getLogger("audio_stream_crypto")

_NONCE_LEN = 12


class AudioDecryptionError(Exception):
    """Raised when an audio chunk fails to decrypt — wrong/rotated key,
    truncated payload, or a tampered/corrupted chunk. Callers must treat
    this the same as "no audio chunk was sent" (see voice_stream.py),
    never crash the whole call-detection session over one bad chunk."""


def decrypt_audio_chunk(payload_b64: str) -> bytes:
    """Decrypts one base64(nonce || ciphertext || tag) chunk. Raises
    AudioDecryptionError on any failure — malformed base64, too-short
    payload, or a failed GCM authentication tag check (wrong key or
    tampered data)."""
    try:
        raw = base64.b64decode(payload_b64)
    except Exception as exc:
        raise AudioDecryptionError(f"invalid base64: {exc}") from exc

    if len(raw) <= _NONCE_LEN:
        raise AudioDecryptionError("payload shorter than the nonce alone — not a real encrypted chunk")

    nonce, ciphertext = raw[:_NONCE_LEN], raw[_NONCE_LEN:]

    try:
        key = base64.b64decode(settings.voice_stream_audio_key_b64)
        aesgcm = AESGCM(key)
        return aesgcm.decrypt(nonce, ciphertext, associated_data=None)
    except InvalidTag as exc:
        raise AudioDecryptionError("GCM authentication failed — wrong key or corrupted/tampered chunk") from exc
    except Exception as exc:
        raise AudioDecryptionError(f"decryption failed: {exc}") from exc

File: apps/mobile/android/app/src/main/java/com/avaran/security/telemetry/AudioStreamCrypto.kt
package com.avaran.security.telemetry

import android.util.Base64
import com.avaran.security.config.DevConfig
import java.security.SecureRandom
import javax.crypto.Cipher
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

/**
 * AES-256-GCM encryption for audio chunks sent to apps/api's
 * /ws/voice-stream — matches app/core/audio_stream_crypto.py's decrypting
 * side byte-for-byte. Raw microphone PCM must never cross the wire in the
 * clear; this is the encrypting half of that guarantee.
 *
 * Wire format: base64(nonce[12 bytes] || ciphertext || tag[16 bytes]) —
 * standard AES-GCM framing, one field, no separate nonce parameter needed
 * on the receiving end.
 */
object AudioStreamCrypto {
    private const val ALGORITHM = "AES/GCM/NoPadding"
    private const val NONCE_LEN_BYTES = 12
    private const val TAG_LEN_BITS = 128

    private val secureRandom = SecureRandom()

    private val key: SecretKeySpec by lazy {
        val keyBytes = Base64.decode(DevConfig.VOICE_STREAM_AUDIO_KEY_B64, Base64.NO_WRAP)
        SecretKeySpec(keyBytes, "AES")
    }

    /** Encrypts one PCM chunk, returning the base64 string ready to send
     * as `audio_chunk_b64`. Never throws on a well-formed key/input — any
     * failure here would silently drop real audio-spoof detection for
     * this chunk, so callers should treat a thrown exception as "skip
     * sending this chunk" rather than crash the detection session. */
    fun encrypt(plainBytes: ByteArray): String {
        val nonce = ByteArray(NONCE_LEN_BYTES)
        secureRandom.nextBytes(nonce)

        val cipher = Cipher.getInstance(ALGORITHM)
        cipher.init(Cipher.ENCRYPT_MODE, key, GCMParameterSpec(TAG_LEN_BITS, nonce))
        val ciphertext = cipher.doFinal(plainBytes)

        val combined = ByteArray(nonce.size + ciphertext.size)
        System.arraycopy(nonce, 0, combined, 0, nonce.size)
        System.arraycopy(ciphertext, 0, combined, nonce.size, ciphertext.size)

        return Base64.encodeToString(combined, Base64.NO_WRAP)
    }
}
